src/backend/distributed/metadata/metadata_sync.c

```diff
@@ -3954,47 +3801,416 @@ ColocationGroupDeleteCommand(uint32 colocationId)
+}
+
+
+/*
+ * EstablishAndSetMetadataSyncBareConnections establishes and sets
+ * connections used throughout nontransactional metadata sync.
+ */
+void
+EstablishAndSetMetadataSyncBareConnections(MetadataSyncContext *context)
+{
+	Assert(MetadataSyncTransMode == METADATA_SYNC_NON_TRANSACTIONAL);
+
+	int connectionFlags = 0;
+
+	/* establish bare connections to activated worker nodes */
+	List *bareConnectionList = NIL;
+	WorkerNode *node = NULL;
+	foreach_ptr(node, context->activatedWorkerNodeList)
+	{
+		MultiConnection *connection = GetNodeUserDatabaseConnection(connectionFlags,
+																	node->workerName,
+																	node->workerPort,
+																	CurrentUserName(),
+																	NULL);
+
+		Assert(connection != NULL);
+		bareConnectionList = lappend(bareConnectionList, connection);
+	}
+
+	context->activatedWorkerConnections = bareConnectionList;
+}
+
+
+/*
+ * EstablishAndSetMetadataSyncCoordinatedConnections establishes and sets
+ * connections used throughout transactional metadata sync.
+ */
+void
+EstablishAndSetMetadataSyncCoordinatedConnections(MetadataSyncContext *context)
+{
+	Assert(MetadataSyncTransMode == METADATA_SYNC_TRANSACTIONAL);
+
+	int connectionFlags = REQUIRE_METADATA_CONNECTION;
+
+	/* establish coordinated connections to activated worker nodes */
+	List *coordinatedConnectionList = NIL;
+	WorkerNode *node = NULL;
+	foreach_ptr(node, context->activatedWorkerNodeList)
+	{
+		MultiConnection *connection =
+			StartNodeConnection(connectionFlags, node->workerName, node->workerPort);
+
+		MarkRemoteTransactionCritical(connection);
+
+		Assert(connection != NULL);
+		coordinatedConnectionList = lappend(coordinatedConnectionList, connection);
+	}
+
+	context->activatedWorkerConnections = coordinatedConnectionList;
+}
+
+
+/*
+ * CreateMetadataSyncContext creates a context which contains worker connections
+ * and a MemoryContext to be used throughout the metadata sync.
+ *
+ * If we collect commands, connections will not be established as caller's intent
+ * is to collcet sync commands.
+ */
+MetadataSyncContext *
+CreateMetadataSyncContext(List *nodeList, bool collectCommands)
+{
+	/* should be alive over multiple transactions during the sync */
+	MemoryContext context = AllocSetContextCreate(TopMemoryContext,
+												  "metadata_sync_context",
+												  ALLOCSET_DEFAULT_SIZES);
+
+	MetadataSyncContext *metadataSyncContext = (MetadataSyncContext *) palloc0(
+		sizeof(MetadataSyncContext));
+
+	metadataSyncContext->context = context;
+	metadataSyncContext->transactionMode = MetadataSyncTransMode;
+	metadataSyncContext->collectCommands = collectCommands;
+	metadataSyncContext->collectedCommands = NIL;
+
+	/* filter the nodes that needs to be activated from given node list */
+	SetMetadataSyncNodesFromNodeList(metadataSyncContext, nodeList);
+
+	/* establish connections */
+	if (!collectCommands && MetadataSyncTransMode == METADATA_SYNC_TRANSACTIONAL)
+	{
```

Comment:
as we discussed in person, do not establish connections in tx mode, let it handled by the transaction/connection management